c10/core/TensorImpl.h

```diff
@@ -1488,8 +1488,10 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    * Compute the number of elements based on the sizes of a tensor.
    */
   int64_t compute_numel() const {
+    static const int64_t largest_int64 = 0x7fffffffffffffff;
     int64_t n = 1;
     for (auto s : sizes()) {
+      TORCH_CHECK(n <= (largest_int64 / s), "int64 overflow while computing numel for sizes ", sizes());
```

Comment:
@ngimel, ever mindful of performance, raises a good question: "do we really want to perform multiple integer divisions every time we create a tensor just to throw a better error message? Integer division is very slow"

Reply:
Yeah division usually takes significantly longer than multiplication. I would like to avoid it, but I'm not sure how.

I know that some architectures (ARM at least) have a status bit that gets flipped if an arithmetic operation overflows. I did some searching to find out if C++ has a portable way of detecting overflows like this, but didn't find it. It still could exist though.

Reply:
I don't think it does because signed integer overflow is undefined behavior.

I guess our question is: is this worth fixing? Seems very corner case, and for the actual issue we would just be upgrading a segfault to an error message which is, admittedly, nicer, but given how niche that scenario is it may not be worth slowing down every PyTorch user.

Reply:
Maybe a better way to detect the overflow would be to cast each element of `sizes()` to uint64_t (they should always be >=0, right?) before doing the multiply. Then, if the result of the unsigned multiply is greater than `0x7fffffffffffffff` or less than either `n` or `s`, then we know that an overflow occurred. Does that make sense? That would be a pretty minimal performance decrease.

Reply:
That's funny you mention that, as @ngimel and I were also discussing that option a moment ago. That would (probably) be faster, true, and that approach is likely the best fix. However it would still be *some* performance impact for a very niche issue. It would have been more natural for PyTorch to represent sizes as unsigned longs, too, but I think there were concerns that math on unsigned longs would take longer than on signed longs (because overflow is defined on unsigned longs). 

Maybe we wait and see if another user hits that issue in a more natural setting?

Reply:
I _think_ the performance impact might be pretty minimal. Since modern CPUs have branch prediction, the next multiply that would have come after an overflown multiply could be executed in (almost) parallel with the overflow detection code. I could be wrong though. Maybe I should do a bit of benchmarking just to see what happens in practice?

Reply:
If you like, sure; it could be fun.

I think you can reduce it to the conversion to unsigned, a check for whether the next value is smaller than the current (or current value is smaller than the previous value, depending on your perspective), and then a check at the end for whether the unsigned -> int conversion will overflow + the actual conversion. 

Of course there are other torch.linalg tasks, too. And not all of them are as significant as torch.linalg.norm ;)

Reply:
I wrote a standalone application to compare three different implementations of `compute_numel`: https://github.com/kurtamohler/pytorch-perf-test-scripts/tree/master/tensor_numel

The three implementations are:
* Original - no overflow detection.
* Overflow divide - implements overflow detection with the division, as this PR currently does.
* Overflow cast - implements overflow detection with an unsigned cast and comparisons, like what we discussed above.

I built the application with the Makefile shown in my link. I used `-O2` since that's what PyTorch's release is built with. If there are any other important flags that might affect performance, let me know and I can update it.

This is essentially what the benchmark does:
```
for ndim in range(21):
    size_vectors = generate_random_size_vectors(num_vectors=1000000, ndim=ndim)
    measure_total_time_of_numel_calls(size_vectors, impl=original)
    measure_total_time_of_numel_calls(size_vectors, impl=overflow_div)
    measure_total_time_of_numel_calls(size_vectors, impl=overflow_cast)
```

The chart below shows the average over 6 different runs of this application:

![Runtime of computing numel with and without overflow detection (average of 6 runs)](https://user-images.githubusercontent.com/8378787/92031394-28fa6b00-ed2e-11ea-83ce-f0d908dac89a.png)

So according to this benchmark, the divide implementation affects performance significantly. The cast-and-compare implementation doesn't appear to affect performance too much in that chart, but here's a speedup comparison against the original:

![Speedup of overflow checking cast-and-compare vs  original (no overflow check) implementation](https://user-images.githubusercontent.com/8378787/92032115-45e36e00-ed2f-11ea-968e-301c95c32116.png)

To me, it seems likely that the ups and downs in this speedup chart are due to noise. What do you think?

Reply:
I can run this on a couple other machines as well, if it would be helpful. The above measurements are from an AMD Ryzen Threadripper 2970WX.

Reply:
Well, I decided to run the benchmark on my laptop and it shows a pretty different story:

![Speedup of overflow checking cast-and-compare vs  original (no overflow check) implementation (Intel i7-9750H)](https://user-images.githubusercontent.com/8378787/92036490-efc5f900-ed35-11ea-998c-b7c6f8b55798.png)

Reply:
This is pretty convincing. Overflow cast it is ;)

Reply:
@mruberry , you might have to reload the page. It doesn't seem to perform the same way for all machines.

Reply:
I did have to reload the page. Would you elaborate on what we're seeing here?

Reply:
I added more details to the comment above.

Reply:
Unfortunately, this method of overflow detection is unreliable. Imagine 3*5 in 3-bit arithmetic. It produces 7 which is larger than 3 and than 5, but it still overflows.

Reply:
Darn, you're right.

Reply:
I _think_ I found a better method:

```
  int64_t n = 1;
  for (auto s : sizes) {
    int64_t n_new = n * s;
    if (s != 0 && n_new / s != n) {
      // Overflow occurred
    }
    n = n_new;
  }
  return n
```

This one still uses a divide, but it does it _after_ the multiply, improving ILP. Doing a couple benchmark runs seems to give pretty similar performance compared to the original implementation. Still I should run more iterations of the benchmark, and I'll write a comprehensive test to make sure this overflow detection method works correctly.

Reply:
That wouldn't work because signed integer overflow is undefined behavior.

I think it's time we stop pursuing this issue. I continue to not think it's worth the time or the risk, especially now that we've looked at a few approaches.

Reply:
Ok makes sense.

Reply:
Sorry @kurtamohler. I'll take the original issue back to triage review to see if we can get it downgraded from high priority, too.